File: src-tauri/src/models/media.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum MediaType {
    Image,
    AnimatedImage,
    Video,
    AvifStatic,
    AvifAnimated,
}

impl MediaType {
    pub fn from_str_value(s: &str) -> Self {
        match s {
            "image" => Self::Image,
            "animated_image" => Self::AnimatedImage,
            "video" => Self::Video,
            "avif_static" => Self::AvifStatic,
            "avif_animated" => Self::AvifAnimated,
            _ => Self::Image,
        }
    }

    pub fn as_str(&self) -> &'static str {
        match self {
            Self::Image => "image",
            Self::AnimatedImage => "animated_image",
            Self::Video => "video",
            Self::AvifStatic => "avif_static",
            Self::AvifAnimated => "avif_animated",
        }
    }
}

/// A single media file within a gallery.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct MediaEntry {
    pub id: i64,
    pub gallery_id: i64,
    pub filename: String,
    pub path: String,
    pub relative_path: String,
    pub sort_order: i64,
    pub group_name: String,
    pub media_type: MediaType,
    pub width: Option<i32>,
    pub height: Option<i32>,
    pub file_size: Option<i64>,
    pub duration_ms: Option<i64>,
    pub is_animated: bool,
    pub mtime: Option<i64>,
}

/// A prefix group extracted from natural sort grouping.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct MediaGroup {
    pub name: String,
    pub start_index: usize,
    pub count: usize,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum ReadingMode {
    Single,
    VerticalScroll,
    DoublePage,
    ThumbnailGrid,
}

/// An unorganized file sitting directly in an artist folder.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct UnorganizedFile {
    pub id: i64,
    pub artist_id: i64,
    pub filename: String,
    pub path: String,
    pub media_type: Option<MediaType>,
    pub file_size: Option<i64>,
}

/// A tag for categorizing galleries.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Tag {
    pub id: i64,
    pub name: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub gallery_count: Option<i64>,
}

File: src/shared/types/media.ts
import type { GallerySortOrder } from './gallery';

export type MediaType = 'image' | 'animated_image' | 'video' | 'avif_static' | 'avif_animated';

export interface MediaEntry {
  id: number;
  galleryId: number;
  filename: string;
  path: string;
  relativePath: string;
  sortOrder: number;
  groupName: string;
  mediaType: MediaType;
  width: number | null;
  height: number | null;
  fileSize: number;
  durationMs: number | null;
  isAnimated: boolean;
  mtime: number;
}

export interface ConvertedMedia {
  originalPath: string;
  convertedPath: string;
  mediaType: 'animated_webp';
}

export interface MediaGroup {
  name: string;
  startIndex: number;
  count: number;
}

export type ReadingMode = 'single' | 'vertical_scroll' | 'double_page' | 'thumbnail_grid' | 'long_strip';
export type FitMode = 'fit_best' | 'fit_width' | 'fit_height' | 'original';
export type ReadingDirection = 'ltr' | 'rtl' | 'vertical';

export interface ReadingState {
  galleryId: number;
  currentPage: number;
  totalPages: number;
  currentGroup: string | null;
  readingMode: ReadingMode;
  zoomLevel: number;
  fitMode: FitMode;
  readingDirection: ReadingDirection;
}

export interface SmartGroup {
  name: string;
  galleryIds: number[];
  galleryNames: string[];
}

export interface ChronologicalGroup {
  date: string;
  galleryId: number;
  galleryName: string;
}

export interface TimelineEntry {
  index: number;
  filename: string;
  date: string | null;
}

export interface VideoPlayerState {
  playing: boolean;
  currentTime: number;
  duration: number;
  volume: number;
  muted: boolean;
  playbackRate: number;
  looping: boolean;
  fullscreen: boolean;
}

export interface Tag {
  id: number;
  name: string;
  galleryCount?: number;
}

export interface UnorganizedFile {
  id: number;
  artistId: number;
  filename: string;
  path: string;
  mediaType: MediaType | null;
  fileSize: number;
}

export interface ScanResult {
  rootId: number;
  artistsFound: number;
  galleriesFound: number;
  mediaFilesFound: number;
  unorganizedFiles: number;
  orphanedZips: number;
  scanDurationMs: number;
  changedFiles: number;
  errors: ScanError[];
}

export interface ScanError {
  path: string;
  message: string;
  severity: 'warning' | 'error';
}

export interface FfmpegStatus {
  available: boolean;
  version: string | null;
  path: string | null;
}

export interface AppSettings {
  theme: 'dark' | 'light';
  language: string;
  defaultReadingMode: ReadingMode;
  autoPlayAnimated: boolean;
  autoPlayLoopThreshold: number;
  thumbnailSize: 'small' | 'medium' | 'large';
  showMediaBadges: boolean;
  videoPlayerVolume: number;
  gallerySortOrder: GallerySortOrder;
  thumbnailCacheMaxMb: number;
  autoExportMetadata: boolean;
  defaultReadingDirection: ReadingDirection;
  defaultFitMode: FitMode;
  autoScrollSpeed: number;
  smartGroupingThreshold: number;
  enableSmartGrouping: boolean;
  enableChronologicalLinking: boolean;
}
